src/rest/rest_utils.c

```diff
@@ -23,11 +23,368 @@
+    {
+        goto exit;
+    }
+
+    if (gnutls_x509_privkey_generate(device_key, GNUTLS_PK_EC, GNUTLS_CURVE_TO_BITS(GNUTLS_ECC_CURVE_SECP256R1), 0))
+    {
+        goto exit;
+    }
+
+    device_entry->serial = malloc(20);
+    if (device_entry->serial == NULL)
+    {
+        goto exit;
+    }
+
+    do
+    {
+        generate_serial(device_entry->serial, &device_entry->serial_len);
+    } while (find_existing_serial(device_entry->serial, device_entry->serial_len, context));
+
+    activation_time = time(NULL);
+
+    if (gnutls_x509_crt_set_version(device_cert, 3)
+        || gnutls_x509_crt_set_serial(device_cert, device_entry->serial, device_entry->serial_len)
+        || gnutls_x509_crt_set_activation_time(device_cert, activation_time)
+        || gnutls_x509_crt_set_expiration_time(device_cert, activation_time + 60 * 60)
+        || gnutls_x509_crt_set_key(device_cert, device_key))
+    {
+        goto exit;
+    }
+
+    if (gnutls_x509_crt_set_subject_alt_name(device_cert, GNUTLS_SAN_DNSNAME, device_entry->uuid, strlen(device_entry->uuid) + 1, GNUTLS_FSAN_SET))
+    {
+        goto exit;
+    }
+
+    if (gnutls_x509_crt_sign(device_cert, ca_cert, ca_key))
+    {
+        goto exit;
+    }
+
+    gnutls_x509_crt_export(device_cert, GNUTLS_X509_FMT_PEM, NULL, &device_entry->public_key_len);
+    gnutls_x509_privkey_export(device_key, GNUTLS_X509_FMT_PEM, NULL, &device_entry->secret_key_len);
+
+    device_entry->public_key = malloc(device_entry->public_key_len);
+    device_entry->secret_key = malloc(device_entry->secret_key_len);
+
+    if (device_entry->public_key == NULL
+        || device_entry->secret_key == NULL)
+    {
+        goto exit;
+    }
+
+    if (gnutls_x509_crt_export(device_cert, GNUTLS_X509_FMT_PEM, device_entry->public_key, &device_entry->public_key_len)
+        || gnutls_x509_privkey_export(device_key, GNUTLS_X509_FMT_PEM, device_entry->secret_key, &device_entry->secret_key_len))
+    {
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    gnutls_free(ca_cert_buffer.data);
+    gnutls_free(ca_key_buffer.data);
+    gnutls_x509_crt_deinit(device_cert);
+    gnutls_x509_privkey_deinit(device_key);
+    gnutls_x509_crt_deinit(ca_cert);
+    gnutls_x509_privkey_deinit(ca_key);
+
+    return ret;
+}
+
+static int device_new_credentials(database_entry_t *device_entry, void *context)
+{
+    if (device_entry->mode == DEVICE_CREDENTIALS_PSK)
+    {
+        if (device_new_psk(device_entry))
+        {
+            return -1;
+        }
+    }
+    else if (device_entry->mode == DEVICE_CREDENTIALS_CERT)
+    {
+        if (device_new_certificate(device_entry, context))
+        {
+            return -1;
+        }
+    }
+    else if (device_entry->mode == DEVICE_CREDENTIALS_NONE)
+    {
+        return 0;
+    }
+    else
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+database_entry_t *database_get_entry_by_name(const char *name, linked_list_t *device_list)
```

Comment:
If function searches for name in device list, then it is database util, if it is database util, database argument should be first (as in other utils, for example linked list).